### PR TITLE
Ignore Loki endpoints in application data bags of breaking relations

### LINKED ISSUE #137: loki_push_api library misbehaves on relation_broken events
### Bug Description

The `loki_push_api` relation providers store data in the application data bag. That data survives all units being deleted (i.e., `juju scape-application loki 0`), and the `loki_push_api` library will provide wrong data with the `endpoints` API over a `relation_broken` hook.

### To Reproduce

Write a unit test that checks what happens when you access the `LokiPushApiConsumer.endpoints` property over a `relation_broken` hook :-)

### Environment

EVERYWHERE

### Relevant log output

```shell
(˚Õ˚)ر ~~~~╚╩╩╝
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mmanciop**
## Issue

When a relation is breaking (`juju remove-relation ...`), we can still access its data bags. This leads to spurious endpoints being reported by `LogPushApiConsumer.loki_endpoints` over `relation_broken` hooks.

Also, since the endpoint data lives in the application data bag of the provider application, the `LokiPushApiConsumer` listening to `relation_departed` provides no value to knowing when endpoints disappear.

## Solution

This commit changes the LokiPushApiConsumer to ignore endpoints found in application data bags of relations that are breaking.

## Context

Closes #137

## Testing Instructions

Try run the unit tests without the changes in the `loki_push_api.py` file and then shake your head in sadness.

## Release Notes

Fix a but in the `LokiPushApiConsumer` that reported invalid Loki endpoints over the hooks for the breaking of the relations that specified them.

### CODE REVIEW

File: lib/charms/loki_k8s/v0/loki_push_api.py [hunk 1]
```diff
@@ -1553,9 +1554,14 @@ def __init__(
         )
         super().__init__(charm, relation_name, alert_rules_path, recursive)
         events = self._charm.on[relation_name]
+
         self.framework.observe(self._charm.on.upgrade_charm, self._on_logging_relation_changed)
         self.framework.observe(events.relation_changed, self._on_logging_relation_changed)
-        self.framework.observe(events.relation_departed, self._on_logging_relation_departed)
+
+        # This will keep track of broken relations so that we can ignore the endpoints
+        # we find in their application data bags
```
Comment (simskij) [inline lib/charms/loki_k8s/v0/loki_push_api.py:1562]:
```suggestion
        # we find in their application data bags
        # will be reinitiated every time an external event
        # (like relation broken) is emitted
```

File: lib/charms/loki_k8s/v0/loki_push_api.py [hunk 2]
```diff
@@ -1553,9 +1554,16 @@ def __init__(
...
+        # will be reinitiated every time an external event
+        # (like relation broken) is emitted
+        self._ignored_relations = []  # type: list
```
Comment (sed-i) [inline lib/charms/loki_k8s/v0/loki_push_api.py:1565]: Interesting! Do we know for sure under what circumstances the charm is not reinitialized, for this to work?

Comment (mmanciop) [reply lib/charms/loki_k8s/v0/loki_push_api.py:1565]: Charm is initialized for every hook coming from Juju. Events that are emitted by a library like this look like recursive calls and do NOT run the charm constructor again.

Comment (mmanciop) [reply lib/charms/loki_k8s/v0/loki_push_api.py:1569]: More info provided in 1b933eafc48a5378e051621b8c7fae397a9b065c

**Comment by mmanciop** [comment]
We should rebase this PR on #139 when the latter merges

File: lib/charms/loki_k8s/v0/loki_push_api.py [hunk 3]
```diff
@@ -1582,9 +1582,14 @@ def __init__(
         )
         super().__init__(charm, relation_name, alert_rules_path, recursive)
         events = self._charm.on[relation_name]
+
         self.framework.observe(self._charm.on.upgrade_charm, self._on_logging_relation_changed)
         self.framework.observe(events.relation_changed, self._on_logging_relation_changed)
-        self.framework.observe(events.relation_departed, self._on_logging_relation_departed)
```
Comment (mmanciop) [inline lib/charms/loki_k8s/v0/loki_push_api.py:1558]: Probably we need ALSO to keep `departed` here to handle downscale.

**Comment by dstathis** [comment]
Closing as this seems to be solved. If anyone disagrees, feel free to reopen.

**Pull request closed without merging**